Slider pedantics

Luminiscental opened the pull request:
Fixes:
- Velocity calculation was wrong on repeat sliders, since `px_len` didn't factor in repeats, but `end_time` did
- The wrong timing points were being applied to objects (off-by-one error)
- The circular arc code broke for certain arrangements / orientations
- Tick timings were wrong when sv was applied (`ms_per_beat` used `difficulty.sm` but ignored the timing point)
- Tick timings were wrong on repeat sliders (the game actually copies the tick positions from the first slide and retrofits timing)
- `time_to_pos` was inaccurate because the points in `gen_points` aren't time-homogenous

Additions:
- Added slider repeats to `tdata`, since they also matter for holding combo (afaik identically to slider ticks)
- Implemented sliderend snapping as in-game so that the length of the curve described by `gen_points` matches `px_len`
- Added the legacy offsets for sliderend leniency and not having slider ticks too close to the sliderend
- For velocity calculation, derive it from the various slider multipliers and then deduce duration, rather than vice versa
- For tick generation, use the beat length from the timing point rather than recalculating it using `px_len` and velocity

I assume this isn't hugely relevant since you guys aren't looking at sliders yet, but I'd just been looking through the game's slider code recently so thought I'd fix this stuff if it could be of use to you.

abraker95 commented:
Thanks for the contribution! I hope to review this pr over the weekend. Definitely do need to test the beatmap processing proper sometime.

abraker95 commented:
For the tests, how are you coming up with slider tick positions? Hovering over them in editor to read out coordinates or some other way?

Luminiscental commented:
> For the tests, how are you coming up with slider tick positions? Hovering over them in editor to read out coordinates or some other way?

Yeah, hover over in editor and if the tests fail with an off-by-one I trusted the results and updated them.
With the 36ms sliderend offset it's trickier cus typically you can't scroll to the exact timestamp, so I'd do an estimate within +-5ms or something and if the result was close to that I updated the test with it.

The whole `assertAlmostEquals(*, places=0)` thing is a bit dubious, might need reworking (I had to update some test with a +1 because of a slider tick timestamp being off by 0.5+ from the rounded value displayed in editor for the corresponding beat; I believe the editor snaps them to beats based on the red line's rounded offset whereas the code here gives values relative to the slider's rounded offset)

abraker95 commented:
The links provided as references to code need to be changed to point to a specific revision instead of master as to avoid linking something that may change in future.

For example,

Link to specific revision: https://github.com/ppy/osu/blob/6944151486e677bfd11f2390163aca9161defbbf/osu.Game/Rulesets/Objects/SliderEventGenerator.cs 

Link to master: 
https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/SliderEventGenerator.cs#L24

abraker95 commented:
Will be using this comment to keep track of what the commits are fixing and possible side effects / concerns since this is a lot to unpack.

1. `bbde83d` (Make slidertick/end generation more accurate)
__About:__ Changes how the last slider tick recorded
====
__Old:__ Last slider tick is recorded at slider's visible end
__New:__ Last slider tick is recorded the way it is in the game
====
__Notes:__ Likely breaks sliders rendering in [Ultimate osu! Analyzer](https://github.com/abraker95/ultimate_osu_analyzer). It depends on last slider tick to be at slider's visible end to know what to draw up to.

abraker95 commented on beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py:
```diff
@@ -31,46 +35,65 @@ def __init__(self, **kargs):
-            return self.gen_points[-1]
+    def __time_to_dist(self, time):
+        percent = value_to_percent(self.start_time(), self.end_time(), time)
+        if percent < 0.0: return 0
+        if percent > 1.0: return self.px_len
```
Why not `percent <= 0.0` and `percent >= 1.0`?

Luminiscental replied:
I initially left [those](https://github.com/abraker-osu/osu_beatmap_reader/blob/efc4a769eac0351896e44209cb2d8b483e3e2648/beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py#L67-L68) checks as-is because I didn't want to change more code than necessary, but I'll remove them since `value_to_percent` uses `bound` anyway.

abraker95 reviewed (changes requested):
Definitely not getting through this in one sitting. Let's start with this.

Luminiscental commented on beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py:
```diff
@@ -31,46 +35,65 @@ def __init__(self, **kargs):
+        # https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/SliderEventGenerator.cs#L79
+        # https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/Legacy/ConvertSlider.cs#L52
+        midpoint_time = (self.start_time() + self.end_time()) / 2
+        end_tick_time = max(self.end_time() - 36, midpoint_time)
```
Ah yeah, and having constants gives a better place to put those links too

Luminiscental commented on beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py:
```diff
@@ -88,27 +111,46 @@ def __process_slider_data(self, sdata):
 
 
+    def __calculate_length_sums(self, px_len, extend):
+        self.length_sums = [ 0 ]
```
Done, also moved the snapping functionality into another function so it's clear what is doing what.

Luminiscental commented on beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py:
```diff
@@ -31,46 +35,65 @@ def __init__(self, **kargs):
-        return x_pos, y_pos
 
+    def __dist_to_pos(self, distance):
+        i = binary_search(self.length_sums, distance)
```
Can do

Luminiscental commented:
Thanks for taking the time to review this! I've addressed some things now so that there's less to keep track of later.

> `bbde83d` ... Notes: Likely breaks sliders rendering in Ultimate osu! Analyzer. It depends on last slider tick to be at slider's visible end to know what to draw up to.

This is an interesting point; if needed the last tick offset can be removed from these changes, I had assumed the rendering relied on the curve data rather than the tick data (visually accurate start/end times are present in `hdata`, and the start/end positions (plus everything in-between) can be obtained from `gen_points` although maybe that's not meant to be exposed).

> `131fa0c` ... Notes: The recorded slider ticks are no longer sorted by time at which they occur

Just for reference this is reverted in a later commit (sorry the history is a bit disorganized, I made some misjudgements initially).

abraker95 commented on beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py:
```diff
@@ -87,29 +133,55 @@ def __process_slider_data(self, sdata):
+
+
+    @staticmethod
+    def __snap_path_length(gen_points, length_sums, px_len, extend):
```
Should be renamed to some like  `__process_snap_path_length`. Needs something to indicate that `gen_points` and `length_sums` are being modified/generated by this function.

Luminiscental replied:
Does `__process_curve_length` seem reasonable? Also for indicating the argument mutation is a doc-comment explaining what it does enough?

abraker95 replied:
Yes, and yes

abraker95 commented on beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py:
```diff
@@ -87,29 +133,55 @@ def __process_slider_data(self, sdata):
+
+
+    @staticmethod
+    def __calculate_length_sums(gen_points):
```
Suggestion to rename to `__get_length_sums` to shorten name

abraker95 commented on beatmap_reader/hitobject/std/std_holdnote_hitobject_base.py:
```diff
@@ -31,46 +60,63 @@ def __init__(self, **kargs):
+
+        # https://github.com/ppy/osu/blob/ed992eed64b30209381f040586b0e8392d1c168e/osu.Game/Rulesets/Objects/SliderEventGenerator.cs#L118-L137
+        for repeat in range(self.repeats):
+            reverse = repeat % 2 == 1
```
I denote intended booleans with `is_` naming prefix; `is_reverse` in this case

abraker95 commented:
> This is an interesting point; if needed the last tick offset can be removed from these changes, I had assumed the rendering relied on the curve data rather than the tick data (visually accurate start/end times are present in hdata, and the start/end positions (plus everything in-between) can be obtained from gen_points although maybe that's not meant to be exposed).

Yea going forward, it's probably best for `gen_points` not to be exposed. There needs to be some overall cleanup regarding what should be exposed and not.

Keeping track of commits (cont.):

6. 77d666f (Implement slider length snapping; The game doesn't require the last anchor of a slider to be at the exact position of the snapped sliderend, and instead truncates or extends accordingly.)
About: If you look in editor and move around the node controlling of slider's end (straight sliders are affected by this the most) you will see that the displayed slider's end snaps around and can extend well beyond the last slider tick.
====
Old: Last recorded point in `gen_points` lies on last slider tick
New: Last recorded point in `gen_points` lies on the calculated slider extension
====
Notes: Instance specific declaration in `__calculate_length_sums` fixed in later commit

7. 360d125 (Fix tick timing on reverses)
About: Fixes to `131fa0c` (Make slider repeats output tick data)

8. 02028ffcff (Fix minor logic errors)
About: Fixes to `93881bc` (Fix slider path parametrization) and `77d666f` (Implement slider length snapping)

9. ab77037 (Add slider stress-test)
About: Adds a map to test against which was found to have 0 ms time interval between slider ends, causing a div-by-zero error when calculating velocity

10. 9a1afdf14f (Fix timing point off-by-one error)
About: Adds assert for case where a hitobject is mapped before any uninherited timing section
====
Notes: Assuming the assert was for debugging or a temporary measure on handling such case. Either way, it's removed in later commit

Luminiscental commented:
> Yea going forward, it's probably best for gen_points not to be exposed. There needs to be some overall cleanup regarding what should be exposed and not.

So regarding this do you want me to remove the end tick offset code? Also I'm confused how this breaks ultimate_osu_analyzer given that it doesn't rely on this repo (it has separate `get_end_time` and `get_last_tick_time` methods and uses a form of `gen_points` for rendering the slider path anyway).

abraker95 commented:
> So regarding this do you want me to remove the end tick offset code? Also I'm confused how this breaks ultimate_osu_analyzer given that it doesn't rely on this repo (it has separate get_end_time and get_last_tick_time methods and uses a form of gen_points for rendering the slider path anyway).

I wasn't sure if it does break ultimate_osu_analyzer but I did remember using generated points to determine what to render. That said, don't worry about that breaking. It was a comment for me to track what side effects the change may have. So long as there is a way to get points forming slider shape, the change is fine.

abraker95 commented:
I am not able to checkout revision following the addition of Stagazer beatmap due to filename exceeding character limit.
```
"C:\Dev\Git\bin\git.exe" checkout -B "slider-pedantics" "Luminiscental/slider-pedantics"
error: cannot stat 'beatmap_reader/unit_tests/maps/osu/DJ Noriken - Stargazer feat. YUC'e (PSYQUI Remix) (Hishiro Chizuru) [Starg-Azer isn't so great Are you kidding me].osu': Filename too long
```

Luminiscental commented:
> I am not able to checkout revision following the addition of Stagazer beatmap due to filename exceeding character limit.

Ah I forgot this was a potential issue with git on windows, I've renamed the file. Since this stuff was getting changed I also updated the file paths to use `os.path.join` so the tests aren't broken on linux.

abraker95 commented:
I have made a test beatmap for sliders. I have attached it as a txt which you can rename to a .osu file. 

I have also went through all slider ticks and notes in that map and have produced coordinates and timings for each one. I created the file by hand, so while it's not in json format, it's close enough. I have marked question marks for entries representing the following:
* `?` - I am not sure, probably fine
* `??` - You should check, it might not be correct
* `???` - Definitely broken

[abraker - unknown (abraker) [slider_test].txt](https://github.com/abraker-osu/osu_beatmap_reader/files/7843781/abraker.-.unknown.abraker.slider_test.txt)
[slider_test_data.txt](https://github.com/abraker-osu/osu_beatmap_reader/files/7843815/slider_test_data.txt)

Luminiscental commented:
> I have also went through all slider ticks and notes in that map and have produced coordinates and timings for each one. I created the file by hand, so while it's not in json format, it's close enough. I have marked question marks for entries representing the following:

Are you sure you ran my branch when doing this? I'm getting different outputs, e.g. for `[1]` the last value in `tdata` I'm getting is `[191, 179, 45714]` whereas your text file has `[ 192, 176, 45750 ]` which the old code returns.

Beyond that, the main other issue I'm seeing is that stable defaults to linear rather than bezier when perfect circle sliders have collinear anchors; I had missed that lazer converted curve types [here](https://github.com/ppy/osu/blob/ed992eed64b30209381f040586b0e8392d1c168e/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs#L318-L322).

Luminiscental commented:
Ok I've fixed the perfect slider issue, and  added the `slider_test` map as a unit test, with comments in the data file on any output that you marked with `?` which hadn't changed with this PR.

I only properly checked those question marked sliders, but from what I saw things seem to match what I see in the editor.

abraker95 commented:
> Based on other data I'm assuming you removed the end tick offset locally or something.

Those values were produced by me manually going into editor and recording coordinates and timing for each tick. Where I saw no tick but saw code giving one I put `?` because I couldn't confirm whether it's correct via editor. I attempted to fill in the last tick for first few cases, but then gave up as sliders got more complex.

abraker95 commented on beatmap_reader/unit_tests/test_beatmap.py:
```diff
@@ -144,6 +145,24 @@ def test_beatmap_loading_weird(self):
+            'beatmap_reader', 'unit_tests', 'data', 'slider_test_data.txt'
+        )
+        with open(data_path, 'r') as f:
+            data = ast.literal_eval(f.read())
```
Executing what is intended to be a data file as a python script is *really* weird. This is supposed to be a json file.

abraker95 commented:
I noticed that the stargazer map takes a bit time to load and traced the issue to the Bezier code, in bezier.py:
```
# Estimate the length of the curve
approx_length = 0

for i in range(len(curve_points) - 1):
     approx_length += dist(curve_points[i], curve_points[i + 1])

# subdivide the curve
ncurve = int(approx_length / Bezier.APPROX_LEVEL) + 2
for i in range(ncurve):
    self.curve_points.append(Bezier.point_at(curve_points, float(i) / float(ncurve - 1)))
```

It slows down for two cases of `curve_points` (at timings 39628 and 85699, respectively) due to calculated curve lengths being quite large. This in turn produces hundreds of curve subdivisions.
```
{
    "0" : [
        [249, 288], [262, 328], [338, 345], [403, 289], 
        [392, 191], [301, 122], [228, 184], [266, 207], 
        [248, 288], [128, 294], [57, 185], [17, 77], 
        [69, -41], [235, -67], [285, -33], [349, 29], 
        [376, 137], [307, 130], [271, 69], [213, 162], 
        [403, 269], [473, 208], [510, 116], [488, 24], 
        [392, -8], [349, 29]
    ],

    "1" : [
        [1338, -349], [-36047, -259088]
    ]
}
```

First slider is render as such
![](https://i.imgur.com/2SsOpB2.png)

Second one as so:
![](https://i.imgur.com/uHvg0oI.png)

Seems to do too much work for what's being displayed. I think the `ncurve` limit can be capped, but it causes some inaccuracy as a result. For example, [this](https://i.imgur.com/OFAo0Mq.png) is due to capping max `ncurves` to 20. What do you think?

I have pushed a commit to `slider_rendering` branch for [osu-play-analyzer](https://github.com/abraker-osu/osu-play-analyzer) if you want to check it out.

Luminiscental commented:
With the `ncurve` issue you bring up, I think using slider velocity should allow a better estimate to be made, e.g. you could do something like
```python
ncurve = min(..., velocity * (end_time - start_time) / PX_PER_CHUNK)
```
at least, and maybe there are smarter ways to estimate the length of time of each segment and use that directly.
 

I had looked at some optimization potential earlier because of how slow the stargazer test was, but I wasn't expecting this PR to get as out of hand as it has, so I didn't want to include it. There were two main things I saw from profiling:

- The distorted sliders where `start_time == end_time` seemed to cause a decent amount of slowdown from processing the curve even though it isn't used anywhere, so moving `__process_curve_points` e.t.c. into `generate_tick_data` after that check gave a decent boost. (this is what's going on with the slider at 85699)
  
- It's not too hard to vectorize `Bezier.point_at` with numpy, so then in `Bezier.__init__` you can do
  ```python
  self.curve_points = Bezier.point_at(curve_points, np.linspace(0, 1, ncurve))
  ```
  which gave a decent speedup. (This along with the `ncurve` point should help with the slider at 39628)


If you want I can implement these changes on this PR.

abraker95 commented:
> I've implemented these and some other optimizations I had looked at earlier here, which I could fast-forward this PR into if you want.

Yes, you can go ahead with that

Luminiscental commented:
For reference here're the amalgamated test times for those commits; the first two have the main impact and it doesn't seem like anything's a regression.
```
HEAD is now at 7542afe Skip curve processing for infinite bpm sliders
  Time (mean ± σ):     693.5 ms ±  20.3 ms    [User: 670.2 ms, System: 16.7 ms]

HEAD is now at 8c70594 Vectorize Bezier.point_at
  Time (mean ± σ):     413.1 ms ±  12.9 ms    [User: 386.1 ms, System: 20.3 ms]
 
HEAD is now at e3679c0 Vectorize __get_length_sums
  Time (mean ± σ):     398.5 ms ±  11.9 ms    [User: 376.6 ms, System: 16.9 ms]
 
HEAD is now at b388b69 Use np.linspace for linear sliders
  Time (mean ± σ):     373.5 ms ±   3.6 ms    [User: 352.7 ms, System: 17.0 ms]
 
HEAD is now at dbb1efc Bound bezier subdivisions using px_len
  Time (mean ± σ):     372.4 ms ±   4.1 ms    [User: 354.4 ms, System: 15.2 ms]
 
HEAD is now at 342ca40 Optimize __time_to_dist
  Time (mean ± σ):     365.7 ms ±   2.0 ms    [User: 347.7 ms, System: 15.4 ms]
 
HEAD is now at 275e6b6 Vectorize __make_circumscribed
  Time (mean ± σ):     356.0 ms ±   5.9 ms    [User: 336.0 ms, System: 15.5 ms]
  ```
